File: python/container_with_most_water.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        height_dict = dict(enumerate(height))
        height_dict = sorted(height_dict.items(), key=lambda x: x[1], reverse=True)

        if height_dict[0][0] < height_dict[1][0]:
            left = 0
            right = 1
        else:
            left = 1
            right = 0

        answer = ((height_dict[right][0] - height_dict[left][0])
                         * min(height_dict[right][1], height_dict[left][1]))
        for i in range(2, len(height_dict)):

            if height_dict[i][0] < height_dict[left][0]:
                left = i
            elif height_dict[i][0] > height_dict[right][0]:
                right = i
            else:
                continue

            answer = max(answer,
                         (height_dict[right][0] - height_dict[left][0])
                         * min(height_dict[right][1], height_dict[left][1])
                         )

        return answer


s = Solution()
print(s.maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]))
print(s.maxArea([1, 1]))
print(s.maxArea([4, 3, 2, 1, 4]))
print(s.maxArea([1, 2, 1]))

print(s.maxArea([0, 2]))

File: scala/ContainerWithMostWater.scala
object ContainerWithMostWater extends App {
  def maxArea(height: Array[Int]): Int = {
    var (l, r) = (0, height.length - 1)
    var answer = 0
    while(l < r) {
      answer = Math.max(answer, (r - l) * Math.min(height(l), height(r)))
      if(height(r) < height(l)) {
        r -= 1
      } else {
        l += 1
      }
    }
    answer
  }

  println(maxArea(Array(1, 8, 6, 2, 5, 4, 8, 3, 7)))
  println(maxArea(Array(1, 1)))
  println(maxArea(Array(1, 2, 2, 1, 2, 3)))
}
